qiskit_nature/second_q/hamiltonians/quadratic_hamiltonian.py

```diff
@@ -123,29 +123,40 @@ def __init__(
                 if not _is_antisymmetric(antisymmetric_part, rtol=rtol, atol=atol):
                     raise ValueError("Antisymmetric part must be antisymmetric.")
 
-        self._hermitian_part = hermitian_part
-        self._antisymmetric_part = antisymmetric_part
-        self._constant = constant
-
-        if self._hermitian_part is None:
-            self._hermitian_part = np.zeros((self._num_modes, self._num_modes))
-        if self._antisymmetric_part is None:
-            self._antisymmetric_part = np.zeros((self._num_modes, self._num_modes))
+        data: dict[str, float | np.ndarray] = {"": constant}
+        if hermitian_part is not None:
+            data["+-"] = hermitian_part
+        if antisymmetric_part is not None:
+            data["++"] = antisymmetric_part
+        super().__init__(data)
+
+    def __repr__(self) -> str:
+        return (
+            f"QuadraticHamiltonian("
+            f"hermitian_part={repr(self.hermitian_part)}, "
+            f"antisymmetric_part={repr(self.antisymmetric_part)}, "
+            f"constant={self.constant}, "
+            f"num_modes={self._num_modes})"
+        )
 
     @property
     def hermitian_part(self) -> np.ndarray:
         """The matrix of coefficients of terms that conserve particle number."""
-        return self._hermitian_part
+        if "+-" in self:
+            return cast(np.ndarray, self["+-"])
+        return np.zeros((self._num_modes, self._num_modes), dtype=complex)
 
     @property
     def antisymmetric_part(self) -> np.ndarray:
         """The matrix of coefficients of terms that do not conserve particle number."""
-        return self._antisymmetric_part
+        if "++" in self:
+            return cast(np.ndarray, self["++"])
+        return np.zeros((self._num_modes, self._num_modes), dtype=complex)
 
     @property
     def constant(self) -> float:
         """The constant."""
-        return self._constant
+        return cast(float, self[""])
 
     @property
     def num_modes(self) -> float:
```

Comment:
Down below in the `second_q_op` method is a `TODO` comment which can be removed.

Also for that method: if you do decide to allow sparse matrices, the `second_q_op` method can be replaced by `FermionicOp.from_polynomial_tensor` which will prevent duplication of the iteration logic.
You would need to register the `"--'` key prior to passing the PT to that method though.

Reply:
It is only necessary if we want to replace the `second_q_op` method with `Fermionic.from_polynomial_tensor`. The benefit of that (as I said) would be immediate support for sparse tensors, too.
This would look somewhat like the following:
```python
def second_q_op(self):
    self._data[""] = constant
    self._data["--"] = self.antisymmetric_part.conjugate()
    return FermionicOp.from_polynomial_tensor(self)
```
It would be better not to modify `self._data` so making a shallow copy of `self._data` and passing a new `PolynomialTensor(shallow_copy_data, copy=False)` into the `FermionicOp.from_polynomial_tensor` method might be better. But this is just to give you the idea

Reply:
What I'm confused about is, why does `FermionicOp.from_polynomial_tensor(self)` not work directly? Adding an extra operator
```
self._data["--"] = self.antisymmetric_part.conjugate()
```
actually cancels out the `++` key. This seems like a bug with `FermionicOp.from_polynomial_tensor` or PolynomialTensor.

Reply:
Thank you. The bug was actually in how I stored the antisymmetric matrix in the first place. I've fixed it.